RUMM-811 Add `rum_collection.md` guide

ncreated opened the pull request:
### What and why?

🗒️ This PR adds `rum_collection.md` guide explaining to users how to use RUM.

### How?

For now, the guide has 3 sections:
* Setup SDK for RUM,
* RUM Auto Instrumentation,
* RUM Manual Instrumentation.

I also made sure this `.md` renders well on documentation website (locally), so we can easily deploy it when we want:

<img width="1215" alt="Screenshot 2020-10-14 at 15 07 21" src="https://user-images.githubusercontent.com/2358722/95993105-f95f7800-0e2e-11eb-9978-dd9c32e6cb88.png">


### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

hdelaby commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,261 @@
+    Global.rum = RUMMonitor.initialize()
+    ```
+
+From here, you have two choices. You can either use our auto instrumentation to let the SDK track RUM Views, Resources, Actions and Errors automatically or you can instrument your app manually to send this data. It is also possible to mix both methods.
````
Just a matter of preference but I'd rather have:

The RUM SDK offers two instrumentation methods:
1. Auto-instrumentation (recommended) - the SDK tracks views, resources, actions and errors automatically
2. Manual instrumentation - you instrument your code to send RUM events

**Note**: It is possible to mix both methods.

what do you think?

ncreated replied:
Yes, it's a better narration, done 👌

hdelaby commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,261 @@
+## Setup
+
+
+1. Declare the library as a dependency depending on your package manager (check our [Releases page][releases] for the recent beta version):
```
```suggestion
1. Declare the library as a dependency depending on your package manager (check our [Releases page][releases] for the latest beta version):
```

ncreated commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,261 @@
+
+### RUM Views
+
+To manually start and stop the RUM View use the `.startView(viewController:)` and `.stopView(viewController:)` methods available on `Global.rum`.
```
✅

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+# iOS RUM Collection
+
+<div class="alert alert-info">The iOS RUM collection is in beta. If you have any questions, contact our <a href="https://docs.datadoghq.com/help/" target="_blank">support team</a>.</div>
```
```suggestion
<div class="alert alert-info">The iOS RUM collection is in beta. If you have any questions, contact the <a href="https://docs.datadoghq.com/help/" target="_blank">support team</a>.</div>
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+Send [Real User Monitoring data][rum] to Datadog from your iOS applications with [Datadog's `dd-sdk-ios` client-side RUM SDK][dd-sdk-ios] and leverage the following features:
+
+* get a global idea about your app’s performance and demographics;
```
```suggestion
* Get a holistic view of your app’s performance and demographics.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+Send [Real User Monitoring data][rum] to Datadog from your iOS applications with [Datadog's `dd-sdk-ios` client-side RUM SDK][dd-sdk-ios] and leverage the following features:
+
+* get a global idea about your app’s performance and demographics;
+* understand which resources are the slowest;
```
```suggestion
* Understand which resources are the slowest.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+* get a global idea about your app’s performance and demographics;
+* understand which resources are the slowest;
+* analyze errors by OS and device type.
```
```suggestion
* Analyze errors by OS and device type.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+## Setup
+
+1. Declare the library as a dependency depending on your package manager (check our [Releases page][releases] for the latest beta version):
```
```suggestion
1. Declare the library as a dependency, depending on your package manager. See Datadog's [Releases page][releases] for the latest beta version.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+    {{% /tab %}}
+    {{% tab "Swift Package Manager (SPM)" %}}
+
+To integrate using Apple's [Swift Package Manager][spm], add the following as a dependency to your `Package.swift`:
```
to integrate what?

ncreated replied:
```suggestion
To integrate the SDK using Apple's [Swift Package Manager][spm], add the following as a dependency to your `Package.swift`:
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+    {{% /tab %}}
+    {{< /tabs >}}
+
+2. Initialize the library with your application context and your [Datadog client token][client-token]. For security reasons, you must use a client token: you cannot use [Datadog API keys][api-keys] to configure the `dd-sdk-ios` library as they would be exposed client-side in the iOS application IPA byte code. For more information about setting up a client token, see the [client token documentatio ...
```
```suggestion
2. Initialize the library with your application context and your [Datadog client token][client-token]. For security reasons, you must use a client token: you cannot use [Datadog API keys][api-keys] to configure the `dd-sdk-ios` library as they would be exposed client-side in the iOS application IPA byte code. For more information about setting up a client token, see the [client token documentation][client-token]. You also need to provide an Application ID (see the [RUM Getting Started page][rum-getting-started]).
```

buranmert commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+   .track(firstPartyHosts: ["your.domain.com"])
+   .build()
+```
+Also, assign our `DDURLSessionDelegate()` as a `delegate` of the `URLSession` you want to monitor, e.g.:
````
maybe we can state that if one of them isn't there, it won't work?

ncreated replied:
IMO we should keep it short. This guide shows how to configure things to make it work, and we should not focus on explaining missconfiguration.

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+The RUM SDK offers two instrumentation methods:
+
+- Auto-instrumentation (recommended) - the SDK tracks views, resources, actions and errors automatically.
```
```suggestion
- Auto-instrumentation (recommended) - the SDK tracks views, resources, actions, and errors automatically.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+**Note**: It is possible to mix both methods.
+
+## Auto Instrumentation
```
```suggestion
## Auto-instrumentation
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+   .build()
+```
+
+The `predicate` must be a class or struct conforming to our `UIKitRUMViewsPredicate` protocol:
````
```suggestion
The `predicate` must be a class or struct conforming to the `UIKitRUMViewsPredicate` protocol:
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+}
+```
+
+Inside the `rumView(for:)` implementation, your app should decide if a given `UIViewController` instance should start the RUM view or not (return `nil` in such case). The returned value of `RUMViewFromPredicate` should specify at least the `path` for created RUM view. Refer to code documentation comments for more details.
````
```suggestion
Inside the `rumView(for:)` implementation, your app should decide if a given `UIViewController` instance should start the RUM view or not (and return `nil` in this case). The returned value of `RUMViewFromPredicate` should specify at least the `path` for created the RUM view. Refer to code documentation comments for more details.
```

can we link to these comments?

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+Inside the `rumView(for:)` implementation, your app should decide if a given `UIViewController` instance should start the RUM view or not (return `nil` in such case). The returned value of `RUMViewFromPredicate` should specify at least the `path` for created RUM view. Refer to code documentation comments for more details.
+
+**Note**: the SDK will call `rumView(for:)` many times while your app is running. Your implementation of the predicate should not depend on the order of SDK calls.
```
```suggestion
**Note**: The SDK calls `rumView(for:)` many times while your app is running. Your implementation of the predicate should not depend on the order of SDK calls.
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+   .track(firstPartyHosts: ["your.domain.com"])
+   .build()
+```
+Also, assign our `DDURLSessionDelegate()` as a `delegate` of the `URLSession` you want to monitor, e.g.:
````
```suggestion
Also, assign `DDURLSessionDelegate()` as a `delegate` of the `URLSession` you want to monitor, for example:
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+)
+```
+
+This will make the SDK track requests sent from this instance of the `URLSession`. Requests which URLs match the `firstPartyHosts` will be additionally marked as "first party" in the RUM Explorer.
````
```suggestion
This makes the SDK track requests sent from this instance of the `URLSession`. Requests whose URLs match the `firstPartyHosts` are additionally marked as "first party" in the RUM Explorer.
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+   .build()
+```
+
+This will make the SDK track all significant taps occurring in the app. For privacy reasons, all interactions with the on-screen keyboard are ignored.
````
```suggestion
This makes the SDK track all significant taps occurring in the app. For privacy reasons, all interactions with the on-screen keyboard are ignored.
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+### RUM Errors
+
+All "error" and "critical" logs will be reported as RUM errors and linked to the current RUM view:
```
```suggestion
All "error" and "critical" logs are be reported as RUM errors and linked to the current RUM view:
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+logger.critical("message")
+```
+
+Similarly, all ended APM spans marked as error will be reported as RUM errors:
````
```suggestion
Similarly, all ended APM spans marked as error are be reported as RUM errors:
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+  Global.rum.stopView(viewController: self)
+}
+```
+For more details and available options, please refer to the code documentation comments.
````
```suggestion
For more details and available options, refer to the code documentation comments.
```
link to comments

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+**Note**: the `String` used for `resourceName` in both calls must be unique for the resource you are calling, so the SDK can match resource start with its completion. 
+
+For more details and available options, please refer to the code documentation comments.
```
```suggestion
For more details and available options, refer to the code documentation comments.
```

link to comments

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+)
+```
+
+**Note**: the `String` used for `resourceName` in both calls must be unique for the resource you are calling, so the SDK can match resource start with its completion. 
````
```suggestion
**Note**: The `String` used for `resourceName` in both calls must be unique for the resource you are calling. This is necessary for the SDK to match a resource's start with its completion. 
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+* `.startUserAction(type:name:)`
+* and `.stopUserAction(type:)`
+
+on `Global.rum`. The first method can be used for sending actions which have no time (e.g. `.tap`), while the other two should be used for actions which define start and stop time (e.g. `.scroll`).
```
```suggestion
on `Global.rum`. The first method can be used for sending actions that have no time (e.g. `.tap`), while the other two are intended for actions that define start and stop time (e.g. `.scroll`).
```

ncreated replied:
This one was changed on request, while editing, sorry 🙂.

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+}
+```
+
+**Note**: when using `.startUserAction(type:name:)` and `.stopUserAction(type:)`, the action `type` must be the same, so the SDK can match the action start with completion.
````
```suggestion
**Note**: when using `.startUserAction(type:name:)` and `.stopUserAction(type:)`. This is necessary for the SDK to match a resource's start with its completion. 
```

cswatt commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+**Note**: when using `.startUserAction(type:name:)` and `.stopUserAction(type:)`, the action `type` must be the same, so the SDK can match the action start with completion.
+
+For more details and available options, please refer to the code documentation comments.
```
^ see above

ncreated replied:
```suggestion
For more details and available options, refer to the code documentation comments.
```

cswatt commented on docs/rum_collection.md:
````diff
@@ -0,0 +1,271 @@
+rumMonitor.addViewError(message: "error message.", source: .source)
+```
+
+For more details and available options, please refer to the code documentation comments.
````
see above

ncreated replied:
```suggestion
For more details and available options, refer to the code documentation comments.
```

cswatt reviewed (changes requested):
feedback

ncreated commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+
+### RUM Actions
+
+To manually register RUM action, use either:
+* `.registerUserAction(type:name:)`
+
+or:
+* `.startUserAction(type:name:)`
+* and `.stopUserAction(type:)`
+
+on `Global.rum`. The first method can be used for sending actions which have no time (e.g. `.tap`), while the other two should be used for actions which define start and stop time (e.g. `.scroll`).
```
👍 but there's a typo. This is correct:

```suggestion
To manually register instantaneous RUM actions (e.g: `.tap`), use:
* `.registerUserAction(type:name:)`

or for continuous RUM actions (e.g: `.scroll`), use:
* `.startUserAction(type:name:)`
* and `.stopUserAction(type:)`

on `Global.rum`.
```

cswatt commented:
Changes look good! Still needs to link to code documentation comments (or at least have more indication of how to find these comments). Also would like for the markdown links to use numbers rather than words, as that's the style used throughout docs.

ncreated commented:
> Changes look good! Still needs to link to code documentation comments (or at least have more indication of how to find these comments). Also would like for the markdown links to use numbers rather than words, as that's the style used throughout docs.

Thank you so much @cswatt 🙂! I updated the links to be numbers. And added more context: 
```diff
+ refer to the code documentation comments in `DDRUMMonitor` class
```
What I mean with this is just "see the documentation we provide for each public method". As we don't store this doc anywhere online, I can't put a true link. I may reference the source file in our repo, but this could quickly become obsolete if we move this file around.

hdelaby commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+[2]: https://github.com/DataDog/dd-sdk-ios
+[3]: https://github.com/DataDog/dd-sdk-ios/releases
+[7]: https://docs.datadoghq.com/account_management/api-app-keys/#client-tokens
+[8]: https://docs.datadoghq.com/real_user_monitoring/installation/?tab=us
```
```suggestion
[8]: https://docs.datadoghq.com/real_user_monitoring/browser/#setup
```

hdelaby commented on docs/rum_collection.md:
```diff
@@ -0,0 +1,271 @@
+    {{% /tab %}}
+    {{< /tabs >}}
+
+2. Initialize the library with your application context and your [Datadog client token][7]. For security reasons, you must use a client token: you cannot use [Datadog API keys][9] to configure the `dd-sdk-ios` library as they would be exposed client-side in the iOS application IPA byte code. For more information about setting up a client token, see the [client token documentation][7]. You also ne ...
```
```suggestion
2. Initialize the library with your application context and your [Datadog client token][7]. For security reasons, you must use a client token: you cannot use [Datadog API keys][9] to configure the `dd-sdk-ios` library as they would be exposed client-side in the iOS application IPA byte code. For more information about setting up a client token, see the [client token documentation][7]. You also need to provide an Application ID (create a Javascript RUM application as explained in the [RUM Getting Started page][8]).
```

hdelaby reviewed (commented):
For the time of the private beta, users cannot select "iOS" when creating their RUM application. It is now clearer.
